pop-api/src/error.rs

```diff
@@ -0,0 +1,255 @@
+	Token(TokenError) = 7,
+	/// An arithmetic error.
+	Arithmetic(ArithmeticError) = 8,
+	/// The number of transactional layers has been reached, or we are not in a transactional
+	/// layer.
+	Transactional(TransactionalError) = 9,
+	/// Resources exhausted, e.g. attempt to read/write data which is too large to manipulate.
+	Exhausted = 10,
+	/// The state is corrupt; this is generally not going to fix itself.
+	Corruption = 11,
+	/// Some resource (e.g. a preimage) is unavailable right now. This might fix itself later.
+	Unavailable = 12,
+	/// Root origin is not allowed.
+	RootNotAllowed = 13,
+	// TODO: make generic and add docs.
+	UseCaseError(FungiblesError) = 254,
+	DecodingFailed = 255,
+}
+
+impl FromStatusCode for PopApiError {
+	fn from_status_code(status_code: u32) -> core::result::Result<(), Self> {
+		match status_code {
+			0 => Ok(()),
+			_ => Err(convert_to_pop_api_error(status_code)),
+		}
+	}
+}
+
+// `pub` because it is used in `runtime/devnet/src/extensions/tests/mod.rs`'s test:
+// `dispatch_error_to_status_code_to_pop_api_error_works`
+//
+// This function converts a given `status_code` (u32) into a `PopApiError`. First it encodes the
+// status code into a 4-byte array and checks for unknown nested errors. If decoding into
+// `PopApiError` fails (e.g. a breaking change in the `DispatchError`), it handles the error by
+// converting it to the `Other` variant by shifting each byte one position forward (the last byte is
+// not used for anything)and setting the first byte to 0. If decoding succeeds, it checks if the
+// error is of the `Module` variant and performs any necessary conversion based on the use case.
+pub fn convert_to_pop_api_error(status_code: u32) -> PopApiError {
+	let mut encoded: [u8; 4] =
+		status_code.encode().try_into().expect("qid u32 always encodes to 4 bytes");
+	encoded = check_for_unknown_nesting(encoded);
+	let error = match PopApiError::decode(&mut &encoded[..]) {
+		Err(_) => {
+			encoded[3] = encoded[2];
+			encoded[2] = encoded[1];
+			encoded[1] = encoded[0];
+			encoded[0] = 0;
+			PopApiError::decode(&mut &encoded[..]).unwrap().into()
+		},
+		Ok(error) => {
+			if let crate::PopApiError::Module { index, error } = error {
+				// TODO: make generic.
+				convert_to_fungibles_error(index, error)
+			} else {
+				error
+			}
+		},
+	};
+	ink::env::debug_println!("PopApiError: {:?}", error);
+	error
+}
+
+// If a unknown nested variant of the `DispatchError` is detected meaning any of the subsequent
+// bytes are non-zero (e.g. breaking change in the DispatchError), the error needs to be converted
+// into `PopApiError::Other`'s encoded value. This conversion is done by shifting the bytes one
+// position forward (the last byte is discarded as it is not being used) and replacing the first
+// byte with the `Other` encoded value (0u8). This ensures that the error is correctly categorized
+// as an `Other` variant.
+fn check_for_unknown_nesting(encoded_error: [u8; 4]) -> [u8; 4] {
```

Comment:
Name isnt great, doesnt imply that that an adjustment is being made. Also, would it not be better to take a mutable reference and adjust in place rather than allocating a new array?

Reply:
Completely agree with both